### PR TITLE
fix: fixed bug in deserialization method

### PR DESCRIPTION
**Pull request by ShadowCurse**
## Changes
Fixed byte access with `strip_prefix` checks + added unit tests with unicode.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: src/vmm/src/cpu_config/templates_serde.rs [hunk 1]
```diff
@@ -24,26 +24,19 @@ macro_rules! deserialize_from_str {
...
+                $type::from_str_radix(s, 16)
             } else {
-                $type::from_str(&number_str).map_err(|err| {
-                    D::Error::custom(format!(
-                        "Failed to parse string [{}] as a decimal number for CPU template - {:?}",
-                        number_str, err
-                    ))
-                })?
-            };
+                $type::from_str(&number_str)
```
Comment (mattschlebusch) [inline src/vmm/src/cpu_config/templates_serde.rs:32]: I think we can remove support for decimal numbers here.

**Review by mattschlebusch — COMMENTED** [review]
Adding a unit test to check that strange characters are handleable is probably worthwhile, to duplicate any deserialization bug that was observed previously, that the new code fixes etc. Something similar to the following should be ok:
```
        // Malformed 64-bit address with invalid string encoding
        let cpu_template_result = serde_json::from_str::<CustomCpuTemplate>(
            r#"{
                    "msr_modifiers":  [
                        {
                            "addr": "xϽ0b0000000000000000000000000000000000000000000",
                            "bitmap": "0b0000000xxxxx11110000"
                        },
                    ]
                }"#,
        );
        assert!(cpu_template_result.is_err());
        assert!(cpu_template_result.unwrap_err().to_string().contains(
            "Failed to parse string [xϽ0b0000000000000000000000000000000000000000000]"
        ));
```

File: src/vmm/src/cpu_config/templates_serde.rs (hunk 1 above)
Comment (ShadowCurse) [reply src/vmm/src/cpu_config/templates_serde.rs:32]: Ok guys, you really don't like this 1 line of code, do you?

Comment (kalyazin) [reply src/vmm/src/cpu_config/templates_serde.rs:32]: The reasons I don't like it are:
- I can't see a real use case for it to justify additional code
- It may catalyse a human error due to a high character overlap between dec and hex (eg `4000` instead of intended `0x4000`).

Comment (ShadowCurse) [reply src/vmm/src/cpu_config/templates_serde.rs:32]: 1. The code size is same (or even smaller) with dec decode, rather then without it, because in the last `else` we need to create a `Err("some error string")`.
2. the more probable error is between "0b..." and "0x...". It is much more obvious when value is dec or not.

Comment (zulinx86) [reply src/vmm/src/cpu_config/templates_serde.rs:32]: What actual use case are you imagining for using decimal values for CPUID leaf / MSR address / Aarch64 register IDs?
Minimizing attach surface gives us safety.

Comment (ShadowCurse) [reply src/vmm/src/cpu_config/templates_serde.rs:32]: Well, initially I was defining aarch64 regs ids as decimals. But overall don't have strong use cases for or against decimal format. It is just feel correct to support all of them. The only reason to drop decimals will be if we would force a specific string length i.e for u32 it must be 32 + 2 chars long in binary and 8 + 2 in hex.

> Minimizing attack surface gives us safety.

If only we would have thought about this every time we add new features...

Comment (ShadowCurse) [reply src/vmm/src/cpu_config/templates_serde.rs:32]: Just tried to remove decimal format format and it seems much more work then it should be. Decimal format is used for flags in all x86 templates. Is it really worth it to replace all 0 with 0x0 just to remove decimal format that does no hurt anyone and is even already used by us?

Comment (mattschlebusch) [reply src/vmm/src/cpu_config/templates_serde.rs:32]: I am ok deferring the decimal support change personally. Not important for the case the PR is addressing.

**Review by mattschlebusch — APPROVED** [review]

**Review by kalyazin — APPROVED** [review]
